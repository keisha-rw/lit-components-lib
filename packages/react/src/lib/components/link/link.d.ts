import * as React from 'react';
export declare const PdsLink: React.FunctionComponent<Partial<{
    ariaCurrent: "page" | "step" | "location" | "date" | "time" | "true";
    ariaLabel: string;
    role: string;
    rel: string | undefined;
    target: "_blank" | "_self" | "_parent" | "_top" | undefined;
    size: "default" | "sm" | "lg" | "xl";
    variant: "inverted" | "default" | "emphasis" | "emphasis-inverted";
    href: string;
    type: string | undefined;
    button: "" | "icon" | "default" | "default-inverted" | "primary" | "primary-inverted" | "icon-inverted";
    download: string | boolean | undefined;
    hreflang: string | undefined;
    hover: boolean;
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{
    onClick: (e: Event) => void;
}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
