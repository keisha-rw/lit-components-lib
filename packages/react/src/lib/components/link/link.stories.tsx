import * as React from 'react';
import { StoryFn } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import {
  PdsIconMoreHorizontal,
  PdsIconCheck,
} from '@keisha/design-system-icons-react';
// @ts-ignore
import { themedefault } from '@keisha/design-system-tokens';
import { PdsLink } from './link';

export default {
  title: 'Components/Link',
  component: PdsLink,
  tags: ['autodocs'],
  parameters: {
    actions: {
      handles: ['pds-link-click'],
    },
  },
};

const Template: StoryFn<typeof PdsLink> = (args) => (
  <PdsLink onClick={action('pds-link-click')} {...args}>
    Link text
  </PdsLink>
);

export const Default = Template.bind({});
Default.args = {
  href: 'https://www.google.com',
};

export const DefaultInverted = Template.bind({});
DefaultInverted.storyName = 'Default inverted';
DefaultInverted.args = {
  variant: 'inverted',
  href: 'https://www.google.com',
};

DefaultInverted.parameters = {
  backgrounds: { default: 'BrandXStrong' },
};

export const Emphasis = Template.bind({});
Emphasis.args = { variant: 'emphasis', href: 'https://www.google.com' };

export const EmphasisInverted = Template.bind({});
EmphasisInverted.storyName = 'Emphasis inverted';
EmphasisInverted.args = {
  variant: 'emphasis-inverted',
  href: 'https://www.google.com',
};
EmphasisInverted.parameters = {
  backgrounds: { default: 'BrandXStrong' },
};

const TemplateLeft: StoryFn<typeof PdsLink> = (args) => (
  <PdsLink {...args}>
    <span slot="icon-left">
      <PdsIconCheck color={themedefault.SemanticBorderInteractiveDefault} />
    </span>
    Link text
  </PdsLink>
);
export const IconLeft = TemplateLeft.bind({});
IconLeft.storyName = 'Icon left';
IconLeft.args = {
  href: 'https://www.google.com',
};

const TemplateRight: StoryFn<typeof PdsLink> = (args) => (
  <PdsLink {...args}>
    Link text
    <span slot="icon-right">
      <PdsIconCheck color={themedefault.SemanticBorderInteractiveDefault} />
    </span>
  </PdsLink>
);
export const IconRight = TemplateRight.bind({});
IconRight.storyName = 'Icon right';
IconRight.args = { href: 'https://www.google.com' };

export const Download = Template.bind({});
Download.args = {
  download: true,
  href: 'https://www.google.com',
};

export const Hover = Template.bind({});
Hover.storyName = 'Hover triggered';
Hover.args = {
  hover: true,
  href: 'https://www.google.com',
};

export const ButtonLinkDefault = Template.bind({});
ButtonLinkDefault.storyName = 'Button link default';
ButtonLinkDefault.args = {
  button: 'default',
  href: 'https://www.google.com',
};

export const ButtonLinkDefaultSmall = Template.bind({});
ButtonLinkDefaultSmall.storyName = 'Button link default small';
ButtonLinkDefaultSmall.args = {
  button: 'default',
  size: 'sm',
  href: 'https://www.google.com',
};

export const ButtonLinkDefaultInverted = Template.bind({});
ButtonLinkDefaultInverted.storyName = 'Button link default inverted';
ButtonLinkDefaultInverted.args = {
  button: 'default-inverted',
  href: 'https://www.google.com',
};
ButtonLinkDefaultInverted.parameters = {
  backgrounds: { default: 'BrandXStrong' },
};

export const ButtonLinkPrimary = Template.bind({});
ButtonLinkPrimary.storyName = 'Button link primary';
ButtonLinkPrimary.args = {
  button: 'primary',
  href: 'https://www.google.com',
};

export const ButtonLinkPrimaryInverted = Template.bind({});
ButtonLinkPrimaryInverted.storyName = 'Button link primary inverted';
ButtonLinkPrimaryInverted.args = {
  button: 'primary-inverted',
  href: 'https://www.google.com',
};
ButtonLinkPrimaryInverted.parameters = {
  backgrounds: { default: 'BrandXStrong' },
};

const ButtonLinkIconTemplate: StoryFn<typeof PdsLink> = (args) => (
  <PdsLink {...args}>
    <PdsIconMoreHorizontal
      color={themedefault.SemanticBorderInteractiveDefault}
    />
  </PdsLink>
);
export const ButtonLinkIcon = ButtonLinkIconTemplate.bind({});
ButtonLinkIcon.storyName = 'Button link icon';
ButtonLinkIcon.args = {
  button: 'icon',
  href: 'https://www.google.com',
};

const ButtonLinkIconInvertedTemplate: StoryFn<typeof PdsLink> = (args) => (
  <PdsLink {...args}>
    <PdsIconMoreHorizontal color={themedefault.SemanticBorderInvertedDefault} />
  </PdsLink>
);

export const ButtonLinkIconInverted = ButtonLinkIconInvertedTemplate.bind({});
ButtonLinkIconInverted.storyName = 'Button link icon inverted';
ButtonLinkIconInverted.args = {
  button: 'icon-inverted',
  href: 'https://www.google.com',
};
ButtonLinkIconInverted.parameters = {
  backgrounds: { default: 'BrandXStrong' },
};
