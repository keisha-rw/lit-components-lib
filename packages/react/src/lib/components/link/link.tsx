import * as React from 'react';
import { EventName } from '@lit-labs/react';
import { PdsLink as PdsLinkElement } from '@keisha/design-system';
import { PdsLinkProps } from '../../built-component-props';
import { createPdsReactComponent } from '../../create-pds-react-component';

export const PdsLink = createPdsReactComponent<PdsLinkProps>()({
  tagName: 'pds-link',
  elementClass: PdsLinkElement,
  react: React,
  events: {
    onClick: 'pds-link-click' as EventName<CustomEvent>,
  },
});
