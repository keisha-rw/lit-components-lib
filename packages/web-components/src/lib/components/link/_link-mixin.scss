/*------------------------------------*\
 #LINK_MIXIN
\*------------------------------------*/

@mixin link {
  @include pds-semantic-label-default;

  color: var(--pds-link-text-alt, var(--pds-semantic-text-interactive-default));
  text-decoration: none;
  border-bottom: none;
  display: inline-flex;
  align-items: center;
  justify-content: center;

  &:hover,
  &.pds-c-link--hover,
  &:active {
    text-decoration: underline;
    text-underline-offset: size(0.75);
    color: var(
      --pds-link-text-alt-hover,
      var(--pds-semantic-text-interactive-xstrong)
    );
  }

  &:focus:not(:disabled) {
    outline: none;
  }

  &:focus-visible:not(:disabled) {
    @include link-focused;
  }
}

@mixin link-focused {
  outline: none;
  position: relative;

  &::before {
    content: '';
    position: absolute;
    inset: size(-0.25);
    border: size(0.25) solid var(--pds-component-button-default-border);
  }
}

@mixin link-inverted {
  color: var(--pds-semantic-text-interactive-inverted-default);

  &:hover,
  &.pds-c-link--hover,
  &:active {
    color: var(--pds-semantic-text-interactive-inverted-xstrong);
  }

  &:focus:not(:disabled) {
    outline: none;
  }

  &:focus-visible:not(:disabled) {
    @include link-inverted-focused;
  }
}

@mixin link-inverted-focused {
  outline: none;
  position: relative;

  &::before {
    content: '';
    position: absolute;
    inset: size(-0.25);
    border: size(0.25) solid
      var(--pds-component-button-default-inverted-active-border);
  }
}

@mixin link-emphasis {
  text-decoration: underline;
  text-decoration-style: dotted;
  text-underline-offset: size(0.75);

  &:hover,
  &.pds-c-link--hover,
  &:active {
    .pds-c-link__text {
      text-decoration: underline;
      text-underline-offset: size(0.75);
    }
  }
}

@mixin link-emphasis-inverted {
  color: var(--pds-semantic-text-interactive-inverted-default);
  text-decoration: underline;
  text-decoration-style: dotted;
  text-underline-offset: size(0.75);

  &:hover,
  &:active {
    @include link-hover-inverted;
  }

  &:hover,
  &.pds-c-link--hover,
  &:active {
    color: var(--pds-semantic-text-interactive-inverted-xstrong);

    .pds-c-link__text {
      text-decoration: underline;
      text-underline-offset: size(0.75);
    }
  }

  &:focus:not(:disabled) {
    outline: none;
  }

  &:focus-visible:not(:disabled) {
    @include link-inverted-focused;
  }
}

@mixin link-hover-inverted {
  color: var(--pds-semantic-text-inverted-xstrong);
}
