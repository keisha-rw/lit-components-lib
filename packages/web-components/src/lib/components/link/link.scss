@import './link-mixin';
@import '../button/button-mixin';

/*------------------------------------*\
 #LINK
\*------------------------------------*/

.pds-c-link {
  @include link;
}

.pds-c-link--inverted {
  @include link-inverted;
}

.pds-c-link--emphasis {
  @include link-emphasis;
}

.pds-c-link--emphasis-inverted {
  @include link-emphasis-inverted;
}

.pds-c-link--hover {
  .pds-c-link--emphasis-inverted & {
    @include link-hover-inverted;
  }
}

.pds-c-link__icon-right-wrapper {
  margin-left: size(1);

  .pds-c-link--button & {
    margin-left: 0;
  }
}

.pds-c-link__icon-left-wrapper {
  margin-right: size(1);

  .pds-c-link--button & {
    margin-right: 0;
  }
}

.pds-c-link--sm {
  @include pds-semantic-label-sm;
}

.pds-c-link--lg {
  @include pds-semantic-label-lg;
}

.pds-c-link--xl {
  @include pds-semantic-label-xl;
}

.pds-c-link--button {
  @include button;
}

.pds-c-link--button.pds-c-link--sm {
  @include button-sm;
}

.pds-c-link--button-primary {
  @include button-primary;
}

.pds-c-link--button-primary-inverted {
  @include button-primary-inverted;
}

.pds-c-link--button-default {
  @include button-default;
}

.pds-c-link--button-default-inverted {
  @include button-default-inverted;
}

.pds-c-link--button-icon {
  @include button-icon;

  &.pds-c-link--sm {
    padding: size(0.75);
  }
}

.pds-c-link--button-icon-inverted {
  @include button-icon-inverted;

  &.pds-c-link--sm {
    padding: size(0.75);
  }
}
