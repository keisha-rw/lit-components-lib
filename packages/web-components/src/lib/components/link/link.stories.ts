import { Meta, StoryObj } from '@storybook/web-components';
import { html, nothing } from 'lit';
import '@keisha/design-system-icons-web/check';
import '@keisha/design-system-icons-web/more-horizontal';
import './link';

export default {
  title: 'Components/Link',
  component: 'pds-link',
  tags: ['autodocs'],
  argTypes: {
    variant: {
      control: 'select',
      options: ['default', 'emphasis', 'inverted', 'emphasis-inverted'],
    },
    size: {
      control: 'radio',
      options: ['default', 'sm', 'lg', 'xl'],
    },
    download: {
      control: 'boolean',
    },
    rel: {
      control: 'text',
    },
    hreflang: {
      control: 'text',
    },
    target: {
      control: 'text',
    },
    type: {
      control: 'text',
    },
    button: {
      control: 'select',
      options: [
        'default',
        'default-inverted',
        'primary',
        'primary-inverted',
        'icon',
        'icon-inverted',
        '',
      ],
    },
  },
  render: (args) =>
    html`<pds-link
      variant="${args.variant || nothing}"
      size="${args.size || nothing}"
      href="${args.href || nothing}"
      download="${args.download || nothing}"
      rel="${args.rel || nothing}"
      hreflang="${args.hreflang || nothing}"
      target="${args.target || nothing}"
      ariaLabel="${args.ariaLabel || nothing}"
      role="${args.role || nothing}"
      type="${args.type || nothing}"
      button="${args.button || nothing}"
      ?hover="${args.hover}"
      ><span>Link text</span></pds-link
    >`,
  parameters: {
    actions: {
      handles: ['pds-link-click'],
    },
  },
} as Meta;

// span element is needed to wrap plain text slots https://github.com/storybookjs/storybook/issues/15963

export const Default: StoryObj = {
  name: 'Default',
  args: {
    variant: 'default',
    size: 'default',
    href: 'https://www.google.com',
  },
};

export const DefaultInverted: StoryObj = {
  name: 'Default inverted',
  args: {
    variant: 'inverted',
    size: 'default',
    href: 'https://www.google.com',
  },
  parameters: {
    backgrounds: {
      default: 'BrandXStrong',
    },
  },
};

export const Emphasis: StoryObj = {
  name: 'Emphasis',
  args: {
    variant: 'emphasis',
    size: 'default',
    href: 'https://www.google.com',
  },
};

export const EmphasisInverted: StoryObj = {
  name: 'Emphasis inverted',
  args: {
    variant: 'emphasis-inverted',
    size: 'default',
    href: 'https://www.google.com',
  },
  parameters: {
    backgrounds: {
      default: 'BrandXStrong',
    },
  },
};

export const IconLeft: StoryObj = {
  name: 'Icon left',
  args: { size: 'default' },
  render: () =>
    html`<pds-link href="https://www.google.com">
      <pds-icon-check slot="icon-left"></pds-icon-check>
      <span>Link text</span></pds-link
    >`,
};

export const IconRight: StoryObj = {
  name: 'Icon right',
  args: { size: 'default' },
  render: () =>
    html`<pds-link href="https://www.google.com">
      <span>Link text</span><pds-icon-check slot="icon-right"></pds-icon-check
    ></pds-link>`,
};

export const Download: StoryObj = {
  name: 'Download',
  args: { download: true, href: 'https://www.google.com' },
};

export const HoverTriggered: StoryObj = {
  name: 'Hover triggered',
  args: { hover: true, href: 'https://www.google.com' },
};

export const ButtonLinkDefault: StoryObj = {
  name: 'Button link default',
  args: {
    button: 'default',
    href: 'https://www.google.com',
  },
};

export const ButtonLinkDefaultSmall: StoryObj = {
  name: 'Button link default small',
  args: {
    button: 'default',
    href: 'https://www.google.com',
    size: 'sm',
  },
};

export const ButtonLinkDefaultInverted: StoryObj = {
  name: 'Button link default inverted',
  args: {
    button: 'default-inverted',
    href: 'https://www.google.com',
  },
  parameters: {
    backgrounds: {
      default: 'BrandXStrong',
    },
    a11y: {
      config: {
        rules: [
          {
            // Color contrast is sufficient - this is a false positive
            //
            // @link https://dequeuniversity.com/rules/axe/4.6/color-contrast
            id: 'color-contrast',
            enabled: false,
          },
        ],
      },
    },
  },
};

export const ButtonLinkPrimary: StoryObj = {
  name: 'Button link primary',
  args: {
    button: 'primary',
    href: 'https://www.google.com',
  },
};

export const ButtonLinkPrimaryInverted: StoryObj = {
  name: 'Button link primary inverted',
  args: {
    button: 'primary-inverted',
    href: 'https://www.google.com',
  },
  parameters: {
    backgrounds: {
      default: 'BrandXStrong',
    },
  },
};

export const ButtonLinkIcon: StoryObj = {
  name: 'Button link icon',
  args: {
    button: 'icon',
    href: 'https://www.google.com',
  },
  render: (args) =>
    html`<pds-link
      href="${args.href}"
      button="${args.button}"
      role="link"
      ariaLabel="More actions"
    >
      <pds-icon-more-horizontal slot="icon-left"></pds-icon-more-horizontal>
    </pds-link>`,
};

export const ButtonLinkIconInverted: StoryObj = {
  name: 'Button link icon inverted',
  args: {
    button: 'icon-inverted',
    href: 'https://www.google.com',
  },
  parameters: {
    backgrounds: {
      default: 'BrandXStrong',
    },
  },
  render: (args) =>
    html`<pds-link
      href="${args.href}"
      button="${args.button}"
      role="link"
      ariaLabel="More actions"
    >
      <pds-icon-more-horizontal slot="icon-right"></pds-icon-more-horizontal>
    </pds-link>`,
};
