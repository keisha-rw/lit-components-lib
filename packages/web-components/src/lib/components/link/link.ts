import { html } from 'lit';
import { property, queryAssignedElements } from 'lit/decorators.js';
import { ifDefined } from 'lit/directives/if-defined.js';
import { pdsCustomElement as customElement } from '../../decorators/pds-custom-element';
import { PdsElement } from '../PdsElement';
import styles from './link.scss?inline';

const size = ['default', 'sm', 'lg', 'xl'] as const;
export type LinkSize = (typeof size)[number];

/**
 * @summary This component renders a styled anchor tag
 *
 * @slot icon-left This slot holds an icon to the left of the link. Icons are restricted to the PDS icon sets only
 * @slot default This slot holds the link contents
 * @slot icon-right This slot holds an icon to the right of the link. Icons are restricted to the PDS icon sets only
 *
 * @fires pds-link-click A custom event dispatched on click
 */
@customElement('pds-link', {
  category: 'component',
  type: 'component',
  state: 'stable',
  styles,
})
export class PdsLink extends PdsElement {
  /**
   * - **default** renders link for basic action
   * - **emphasis** provide more affordance
   * - **inverted** used on a darker background
   * - **emphasis-inverted** more affordance on a darker background
   */
  @property()
  variant: 'default' | 'emphasis' | 'inverted' | 'emphasis-inverted' =
    'default';

  /**
   * - **default** renders default size
   * - **sm** smaller size
   * - **lg** larger size (not valid for link button)
   * - **xl** extra large size (not valid for link button)
   */
  @property()
  size: LinkSize = 'default';

  /**
   * Redirect url
   */
  @property({ reflect: true })
  href: string;

  /**
   * Indicates that the link references a file to download, see https://www.w3schools.com/tags/att_a_download.asp
   */
  @property()
  download?: string | boolean;

  /**
   * Specifies information about a linked document
   * Automatically set to 'noopener noreferrer' when target is '_blank'
   */
  @property()
  rel?: string;

  /**
   * Specifies the two-character language code of the document in the link
   */
  @property()
  hreflang?: string;

  /**
   * Specifes target to open the linked document
   */
  @property()
  target?: '_self' | '_blank' | '_parent' | '_top';

  /**
   * Specifies an aria-label for the link
   */
  @property()
  ariaLabel: string;

  /**
   * Specifies an aria-current for the link
   */
  @property()
  ariaCurrent: 'page' | 'step' | 'location' | 'date' | 'time' | 'true';

  /**
   * Specifies a role for the link
   */
  @property()
  role: string;

  /**
   * Specifies the Internet media type (formerly known as MIME type)
   * of the linked document for links, or the type of button if button
   * property is true
   */
  @property()
  type?: string | 'button' | 'reset' | 'submit';

  /**
   * Render the link as a button variant
   *
   * - **default** renders the button used for the most common calls to action that don't require as much visual attention.
   * - **default-inverted** renders a default button for use on darker backgrounds.
   * - **primary** renders the button used for the most important calls to action.
   * - **primary-inverted** renders a primary button for use on darker backgrounds.
   * - **icon** renders the button used for icon.
   * - **icon-inverted** renders the button for icons used on darker backgrounds.
   */
  @property()
  button:
    | 'default'
    | 'default-inverted'
    | 'primary'
    | 'primary-inverted'
    | 'icon'
    | 'icon-inverted'
    | '' = '';

  /**
   * Programatically indicate a link should display as hover state
   */
  @property({ type: Boolean })
  hover: boolean = false;

  handleClick(e: MouseEvent) {
    const customEvent = new CustomEvent('pds-link-click', {
      bubbles: true,
      composed: true,
      cancelable: true,
      detail: {
        summary: this.textContent,
      },
    });

    this.dispatchEvent(customEvent);

    if (customEvent.defaultPrevented) {
      e.preventDefault();
    }
  }

  handleSlotChange(e: Event) {
    this.handleSlotValidation(e);
    this.addSizeToLinkIcon();
  }

  /**
   * @internal
   */
  get classNames() {
    return {
      [this.variant]: !!this.variant,
      'icon-left': !!this.slotNotEmpty('icon-left'),
      'icon-right': !!this.slotNotEmpty('icon-right'),
      [this.size]: !!this.size,
      [`button-${this.button}`]: !!this.button,
      button: !!this.button,
      hover: !!this.hover,
    };
  }

  /**
   * This grabs the iconLeft slot
   * @internal
   */
  @queryAssignedElements({ slot: 'icon-left' })
  iconLeftList: HTMLElement[];

  /**
   * This grabs the iconRight slot
   * @internal
   */
  @queryAssignedElements({ slot: 'icon-right' })
  iconRightList: HTMLElement[];

  addSizeToLinkIcon() {
    const icons = [...this.iconLeftList, ...this.iconRightList];
    if (icons && icons.length && size.includes(this.size)) {
      icons.forEach((icon) =>
        icon.setAttribute(
          'size',
          this.size && this.size === 'xl' ? 'lg' : this.size,
        ),
      );
    }
  }

  updated() {
    this.addSizeToLinkIcon();
  }

  render() {
    if (this.target === '_blank') {
      this.rel = 'noopener noreferrer';
    }

    return html`<a
      class="${this.getClass()}"
      href=${ifDefined(this.href)}
      rel=${ifDefined(this.rel)}
      target="${ifDefined(this.target)}"
      part="link"
      download=${ifDefined(this.download)}
      aria-label=${ifDefined(this.ariaLabel)}
      aria-current=${ifDefined(this.ariaCurrent)}
      role=${ifDefined(this.role)}
      type=${ifDefined(this.type)}
      hreflang=${ifDefined(this.hreflang)}
      @click=${this.handleClick}
      >${this.slotNotEmpty('icon-left')
        ? html`<span class="pds-c-link__icon-left-wrapper">
            <slot
              allowed-elements="pds-icon, span"
              name="icon-left"
              @slotchange=${this.handleSlotChange}
            ></slot>
          </span>`
        : ''}
      <span class="pds-c-link__text" part="link-text"><slot></slot></span
      >${this.slotNotEmpty('icon-right')
        ? html`<span class="pds-c-link__icon-right-wrapper">
            <slot
              allowed-elements="pds-icon, span"
              name="icon-right"
              @slotchange=${this.handleSlotChange}
            ></slot>
          </span>`
        : ''}</a
    >`;
  }
}
