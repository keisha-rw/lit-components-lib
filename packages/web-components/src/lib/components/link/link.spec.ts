import * as path from 'path';
import initStoryshots from '@storybook/addon-storyshots';
import { assert, fixture } from '@open-wc/testing';
import { PdsLink } from './link';

import createSnapshots from '../../../../test/utils/snapshots';

initStoryshots({
  storyKindRegex: /Link/,
  framework: 'web-components',
  configPath: path.join(__dirname, '../../../../.storybook'),
  asyncJest: true,
  test: createSnapshots,
});

describe('Link unit tests', () => {
  let element: PdsLink;

  beforeEach(async () => {
    element = await fixture(
      `<pds-link target="_blank" href="https://www.google.com"><span>This is a link</span></pds-link>`,
    );
  });

  it('is an instance of PdsLink', () => {
    expect(element).toBeInstanceOf(PdsLink);
  });

  it('dispatches a custom event on click and gets correct summary', async () => {
    const linkEl = element.shadowRoot?.querySelector('a');

    const handler = jest.fn();
    element.addEventListener('pds-link-click', handler);

    linkEl?.dispatchEvent(new Event('click'));

    return Promise.resolve().then(() => {
      expect(handler.mock.calls[0][0].detail.summary).toEqual('This is a link');
    });
  });

  it('should preventDefault on the original event if the dispatched event is cancelled', async () => {
    const mockEvent = {
      preventDefault: jest.fn(),
    };
    element.addEventListener('pds-link-click', (e) => e.preventDefault());
    // @ts-ignore testing private method
    element.handleClick(mockEvent);

    expect(mockEvent.preventDefault).toBeCalled();
  });

  it('is accessible', async () => {
    await assert.isAccessible(element);
  });

  it('populates the icon-left slot when provided with a PDS icon element', async () => {
    // In this case, the snapshot renders object info for the custom element rather than the actual markup,
    // so we are testing instead that no console error displays (i.e. the slotted element is valid)
    const consoleError = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});

    element = await fixture(
      '<pds-link><pds-icon-check slot="icon-left"></pds-icon-check></pds-link>',
    );

    expect(consoleError).not.toHaveBeenCalled();
    consoleError.mockRestore();
  });

  it('populates the icon-right slot when provided with a span element', async () => {
    element = await fixture(
      '<pds-link><span slot="icon-right"></span></pds-link>',
    );

    expect(element).toMatchSnapshot();
  });

  it('does not populate an icon slot if the element is neither a PDS icon nor a span', async () => {
    const consoleError = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});

    element = await fixture(
      '<pds-link><div slot="icon-left"> An icon would go here</div></pds-link>',
    );

    expect(element).toMatchSnapshot();
    expect(consoleError).toHaveBeenCalledTimes(1);
    consoleError.mockRestore();
  });

  it('populates the icon-left slot in link button when provided', async () => {
    element = await fixture(
      '<pds-link button><span slot="icon-left"> An icon would go here</span></pds-link>',
    );
    expect(element).toMatchSnapshot();
  });

  it('populates the icon-right slot in link button when provided', async () => {
    element = await fixture(
      '<pds-link button><span slot="icon-right"> An icon would go here</span></pds-link>',
    );
    expect(element).toMatchSnapshot();
  });

  it('Passes size down to the icon', async () => {
    element = await fixture(
      '<pds-link><pds-icon-check slot="icon-left"></pds-icon-check>Icon link</pds-link>',
    );

    const iconSlot = element.querySelector('pds-icon-check') as HTMLElement;
    expect(iconSlot.getAttribute('size')).toEqual('default');

    element = await fixture(
      '<pds-link size="sm"><pds-icon-check slot="icon-left"></pds-icon-check>Icon link</pds-link>',
    );

    const smIconSlot = element.querySelector('pds-icon-check') as HTMLElement;
    expect(smIconSlot.getAttribute('size')).toEqual('sm');

    // If the link is xl, expect the icon to be lg instead of xl. This is the only size that should be different
    element = await fixture(
      '<pds-link size="xl"><pds-icon-check slot="icon-left"></pds-icon-check>Icon link</pds-link>',
    );

    const xlIconSlot = element.querySelector('pds-icon-check') as HTMLElement;
    expect(xlIconSlot.getAttribute('size')).toEqual('lg');
  });
});
