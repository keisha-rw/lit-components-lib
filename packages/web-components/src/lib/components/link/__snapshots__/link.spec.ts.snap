// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link unit tests does not populate an icon slot if the element is neither a PDS icon nor a span 1`] = `<pds-link />`;

exports[`Link unit tests populates the icon-left slot in link button when provided 1`] = `
<pds-link
  button=""
>
  <span
    size="default"
    slot="icon-left"
  >
     An icon would go here
  </span>
</pds-link>
`;

exports[`Link unit tests populates the icon-right slot in link button when provided 1`] = `
<pds-link
  button=""
>
  <span
    size="default"
    slot="icon-right"
  >
     An icon would go here
  </span>
</pds-link>
`;

exports[`Link unit tests populates the icon-right slot when provided with a span element 1`] = `
<pds-link>
  <span
    size="default"
    slot="icon-right"
  />
</pds-link>
`;
